src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala

```diff
@@ -160,33 +140,33 @@ class GoogleExtensions(
     googlePubSubDAO.publishMessages(googleServicesConfig.groupSyncTopic, Seq(id.toJson.compactPrint))
 
   override def onGroupUpdate(groupIdentities: Seq[WorkbenchGroupIdentity]): Future[Unit] =
-    onGroupUpdateRecursive(groupIdentities, Seq.empty)
+    onGroupUpdateRecursive(groupIdentities, Seq.empty).unsafeToFuture()
 
-  private def onGroupUpdateRecursive(groupIdentities: Seq[WorkbenchGroupIdentity], visitedGroups: Seq[WorkbenchGroupIdentity]): Future[Unit] =
+  private def onGroupUpdateRecursive(groupIdentities: Seq[WorkbenchGroupIdentity], visitedGroups: Seq[WorkbenchGroupIdentity]): IO[Unit] =
     for {
-      idsAndSyncDates <- Future.traverse(groupIdentities) { id =>
+      idsAndSyncDates <- groupIdentities.toList.traverse { id =>
         directoryDAO.getSynchronizedDate(id).map(dateOption => id -> dateOption)
       }
       // only sync groups that have already been synchronized
       messagesForIdsWithSyncDates = idsAndSyncDates.collect {
         case (gn: WorkbenchGroupName, Some(_)) => gn.toJson.compactPrint
         case (rpn: FullyQualifiedPolicyId, Some(_)) => rpn.toJson.compactPrint
       }
-      _ <- googlePubSubDAO.publishMessages(googleServicesConfig.groupSyncTopic, messagesForIdsWithSyncDates)
-      ancestorGroups <- Future.traverse(groupIdentities) { id =>
-        directoryDAO.listAncestorGroups(id).unsafeToFuture()
+      _ <- IO.fromFuture(IO(googlePubSubDAO.publishMessages(googleServicesConfig.groupSyncTopic, messagesForIdsWithSyncDates)))
```

Comment:
If this future runs things in its own threadpool, we may need to add another line to shift back to main ec (otherwise we may end up running things on the pubsub threadpool unintendedly)

Reply:
If it’s using global execution context, then we should be fine (as long as it’s not blocking)

https://github.com/typelevel/cats-effect/issues/444 is what I’m referring to.

I meant to revisit all ‘fromFuture’ calls, but haven’t got to it yet

Reply:
this is a discussion for another time but after reading that issue I feel like all these execution contexts and context shifts are more trouble than they are worth. I long for the days of single thread per api call. So easy to understand, so easy to debug.

I agree we should get rid of fromFuture but to be complete we need to update our google daos too.